sklearn/partial_dependence.py

```diff
@@ -0,0 +1,574 @@
+        The partial dependence function evaluated on the ``grid``.
+        For regression and binary classification ``n_classes==1``.
+    """
+    n_samples = X.shape[0]
+    pdp = []
+    for row in range(grid.shape[0]):
+        X_eval = X.copy()
+        for i, variable in enumerate(target_variables):
+            X_eval[:, variable] = np.repeat(grid[row, i], n_samples)
+        if isinstance(est, RegressorMixin):
+            try:
+                pdp.append(np.mean(est.predict(X_eval)))
+            except:
+                raise ValueError('Call %s.fit before partial_dependence' %
+                                 est.__class__.__name__)
+        elif isinstance(est, ClassifierMixin):
+            try:
+                pdp_row = est.predict_proba(X_eval)
+            except:
+                raise ValueError('Call %s.fit before partial_dependence' %
+                                 est.__class__.__name__)
+            pdp_row = np.log(np.clip(pdp_row, 1e-16, 1))
+            pdp_row = np.subtract(pdp_row,
+                                  np.mean(pdp_row, 1)[:, np.newaxis])
+            pdp.append(np.mean(pdp_row, 0))
+        else:
+            raise ValueError('est must be a fitted regressor or classifier '
+                             'model.')
+    pdp = np.array(pdp).transpose()
+    if pdp.shape[0] == 2:
+        # Binary classification
+        pdp = pdp[1, :][np.newaxis]
+    elif len(pdp.shape) == 1:
+        # Regression
+        pdp = pdp[np.newaxis]
+    return pdp
+
+
+def _estimated_partial_dependence(est, target_variables, grid, X):
+    """Calculate the partial dependence of ``target_variables``.
+
+    The function will be calculated by calling the ``predict_proba`` method of
+    ``est`` for classification or ``predict`` for regression on the mean of
+    ``X``.
+
+    Parameters
+    ----------
+    est : BaseEstimator
+        A fitted classification or regression model.
+    target_variables : array-like, dtype=int
+        The target features for which the partial dependency should be
+        computed (size should be smaller than 3 for visual renderings).
+    grid : array-like, shape=(n_points, len(target_variables))
+        The grid of ``target_variables`` values for which the
+        partial dependency should be evaluated (either ``grid`` or ``X``
+        must be specified).
+    X : array-like, shape=(n_samples, n_features)
+        The data on which ``est`` was trained.
+
+    Returns
+    -------
+    pdp : array, shape=(n_classes, n_points)
+        The partial dependence function evaluated on the ``grid``.
+        For regression and binary classification ``n_classes==1``.
+    """
+    n_samples = grid.shape[0]
+    X_eval = np.tile(X.mean(0), [n_samples, 1])
+    for i, variable in enumerate(target_variables):
+        X_eval[:, variable] = grid[:, i]
+    if isinstance(est, RegressorMixin):
+        try:
+            pdp = est.predict(X_eval)
+        except:
```

Comment:
A lot of this is just place holders @jnothman :-) I need to figure out what errors to let the estimator pass through and what to explicitly catch in this function. Not ready for code reviews yet. Just structural stuff for now please.